tests/conftest.py

```diff
@@ -143,3 +153,20 @@ def generate_schema_call_counter(*args: Any, **kwargs: Any) -> Any:
 
     monkeypatch.setattr(GenerateSchema, 'generate_schema', generate_schema_call_counter)
     return counter
+
+
+@pytest.fixture(scope='function', autouse=True)
+def validate_json_schemas(monkeypatch: pytest.MonkeyPatch, request: pytest.FixtureRequest) -> None:
+    orig_generate = GenerateJsonSchema.generate
+
+    def generate(*args: Any, **kwargs: Any) -> Any:
+        json_schema = orig_generate(*args, **kwargs)
+        if not request.node.get_closest_marker('skip_json_schema_validation'):
+            try:
+                Draft202012Validator.check_schema(json_schema)
+            except SchemaError:
+                pytest.fail('Failed to validate the JSON Schema against the Draft 2020-12 spec')
```

Comment:
Perhaps this could be a bit more verbose:

1. Can we include the `SchemaError` information in this result?
2. Could we add a note about the fact that this is purely a testing feature, not a runtime `pydantic` check (at this point)?

Perhaps this is a bit excessive, but can we test this test? As in, run a test within a test, and check that this is raised for invalid schema? If not, no worries, but could you document an example of a failing test on this PR just to have as a reference for the blame later?

Reply:
> 1. Can we include the `SchemaError` information in this result?

Pytest will display the exception by default. This is the output you would get for a failing test:

<details>

```
tests/test_json_schema.py F                                                                                                                       [100%]
tests/test_json_schema.py:6476 test_fails - Failed: Failed to validate the JSON Schema against the Draft 2020-12 spec…                            [100%]
======================================================================= FAILURES ========================================================================
______________________________________________________________________ test_fails _______________________________________________________________________

args = (<pydantic.json_schema.GenerateJsonSchema object at 0x7b80723f5820>, {'metadata': {'pydantic_js_annotation_functions':...onSchema.__get_pydantic_json_schema__ of WithJsonSchema(json_schema={'type': 'invalid'}, mode=None)>]}, 'type': 'int'})
kwargs = {'mode': 'validation'}, json_schema = {'type': 'invalid'}

    def generate(*args: Any, **kwargs: Any) -> Any:
        json_schema = orig_generate(*args, **kwargs)
        if not request.node.get_closest_marker('skip_json_schema_validation'):
            try:
>               Draft202012Validator.check_schema(json_schema)

tests/conftest.py:166: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'jsonschema.validators.Draft202012Validator'>, schema = {'type': 'invalid'}
format_checker = <FormatChecker checkers=['date', 'email', 'idn-email', 'idn-hostname', 'ipv4', 'ipv6', 'regex', 'uuid']>

    @classmethod
    def check_schema(cls, schema, format_checker=_UNSET):
        Validator = validator_for(cls.META_SCHEMA, default=cls)
        if format_checker is _UNSET:
            format_checker = Validator.FORMAT_CHECKER
        validator = Validator(
            schema=cls.META_SCHEMA,
            format_checker=format_checker,
        )
        for error in validator.iter_errors(schema):
>           raise exceptions.SchemaError.create_from(error)
E           jsonschema.exceptions.SchemaError: 'invalid' is not valid under any of the given schemas
E           
E           Failed validating 'anyOf' in metaschema['allOf'][3]['properties']['type']:
E               {'anyOf': [{'$ref': '#/$defs/simpleTypes'},
E                          {'type': 'array',
E                           'items': {'$ref': '#/$defs/simpleTypes'},
E                           'minItems': 1,
E                           'uniqueItems': True}]}
E           
E           On schema['type']:
E               'invalid'

../../.pyenv/versions/3.12.4/envs/pydanticdev/lib/python3.12/site-packages/jsonschema/validators.py:317: SchemaError

During handling of the above exception, another exception occurred:

    def test_fails():
>       TypeAdapter(Annotated[int, WithJsonSchema({'type': 'invalid'})]).json_schema()

tests/test_json_schema.py:6478: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/type_adapter.py:142: in wrapped
    return func(self, *args, **kwargs)
pydantic/type_adapter.py:549: in json_schema
    return schema_generator_instance.generate(self.core_schema, mode=mode)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<pydantic.json_schema.GenerateJsonSchema object at 0x7b80723f5820>, {'metadata': {'pydantic_js_annotation_functions':...onSchema.__get_pydantic_json_schema__ of WithJsonSchema(json_schema={'type': 'invalid'}, mode=None)>]}, 'type': 'int'})
kwargs = {'mode': 'validation'}, json_schema = {'type': 'invalid'}

    def generate(*args: Any, **kwargs: Any) -> Any:
        json_schema = orig_generate(*args, **kwargs)
        if not request.node.get_closest_marker('skip_json_schema_validation'):
            try:
                Draft202012Validator.check_schema(json_schema)
            except SchemaError:
>               pytest.fail('Failed to validate the JSON Schema against the Draft 2020-12 spec')
E               Failed: Failed to validate the JSON Schema against the Draft 2020-12 spec

tests/conftest.py:168: Failed
                                   Summary of Failures                                    
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃  File                       ┃  Function    ┃  Function Line  ┃  Error Line  ┃  Error   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│  tests/test_json_schema.py  │  test_fails  │  6477           │  6478        │  Failed  │
└─────────────────────────────┴──────────────┴─────────────────┴──────────────┴──────────┘
Results (1.65s):
         1 failed
      5846 deselected

```

</details>

> 2\. Could we add a note about the fact that this is purely a testing feature, not a runtime `pydantic` check (at this point)?

Added.

> Perhaps this is a bit excessive, but can we test this test?

Seems like there are [ways to do so](https://stackoverflow.com/a/56635224), but they are pretty involved. I added a test with an expected failure.